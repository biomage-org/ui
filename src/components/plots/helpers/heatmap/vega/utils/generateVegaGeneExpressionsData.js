const cartesian = (...array) => (
  array.reduce((acum, value) => (
    acum.flatMap((d) => (
      value.map((e) => [d, e].flat())
    ))
  ))
);

const generateVegaGeneExpressionsData = (cellOrder, geneOrder, expression, heatmapSettings) => {
  const { expressionValue, truncatedValues } = heatmapSettings;

  const geneExpressionsData = [];

  if (!expression.matrix.genesAreLoaded(geneOrder)) {
    return;
  }

  // Preload all genes so that their arrays are generated only once
  const preloadedExpressions = {};
  geneOrder.forEach((gene) => {
    const geneExpression = { rawExpression: expression.matrix.getRawExpression(gene, cellOrder) };
    geneExpression.truncatedExpression = expression.matrix.getTruncatedExpression(gene, cellOrder);

    preloadedExpressions[gene] = geneExpression;
  });

  const cellOrderWithIndexes = cellOrder.map((cellId, index) => ({ cellId, index }));

  cartesian(geneOrder, cellOrderWithIndexes).forEach(
    ([gene, { cellId, index }]) => {
      const expressionValues = {};

      if (expressionValue === 'zScore') {
        throw new Error('Not implemented DO NOT MERGE: Zscore needs to be implemented for sparse matrix in the worker first');
        // expressionValues = {
        //   color: expressionDataForGene.zScore, display: expressionDataForGene.zScore,
        // };
      } else {
        expressionValues.display = preloadedExpressions[gene].rawExpression;
        expressionValues.color = truncatedValues
          ? preloadedExpressions[gene].truncatedExpression
          : preloadedExpressions[gene].display;
      }

      geneExpressionsData.push({
        cellId,
        gene,
        expression: expressionValues.color[index],
        displayExpression: expressionValues.display[index],
      });
    },
  );

  return geneExpressionsData;
};

export default generateVegaGeneExpressionsData;
