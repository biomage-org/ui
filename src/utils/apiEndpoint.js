const getApiEndpoint = (location) => {
  try {
    const url = new URL(location || window.location.href);

    if (url.hostname.includes('staging')) {
      return url.origin.replace('ui', 'api');
    }

    if (url.hostname.includes('localhost') || url.hostname.includes('127.0.0.1')) {
      return 'http://localhost:3000';
    }

    return `https://api.${process.env.DOMAIN_NAME}`;
  } catch (error) {
    console.error('Failed to get API endpoint.');
    console.error(error);
  }
};

export default getApiEndpoint;
