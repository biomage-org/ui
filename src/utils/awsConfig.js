import * as AWS from '@aws-sdk/client-sts';
import { getDefaultRoleAssumerWithWebIdentity } from '@aws-sdk/client-sts';
import { fromTokenFile } from '@aws-sdk/credential-provider-web-identity';

const getAccountId = async () => {
  const sts = new AWS.STS({
    region: getAWSRegion(),
    credentials: fromTokenFile({
      roleAssumerWithWebIdentity: getDefaultRoleAssumerWithWebIdentity(),
    }),
  });
  console.log('HERE LMAO', getAWSRegion());

  const accountID = await sts.getCallerIdentity({}).promise;
  console.log('ID IS ', accountID);

  return accountID;
};

const getAWSRegion = () => {
  const region = process.env.AWS_REGION || process.env.DEFAULT_REGION || 'eu-west-1';
  return region;
};

export { getAccountId, getAWSRegion };
