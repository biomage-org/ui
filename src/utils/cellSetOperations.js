const unionByCellClass = (cellClassKey, hierarchy, properties) => {
  const cellSetKeys = hierarchy
    .find(({ key }) => key === cellClassKey).children
    .map(({ key }) => key);

  return union(cellSetKeys, properties);
};

const union = (listOfSets, properties) => {
  if (!listOfSets) {
    return new Set();
  }
  const sets = listOfSets.map((key) => properties[key]?.cellIds || []);
  const unionSet = new Set(
    [].concat(
      ...sets.map(
        (set) => [...set],
      ),
    ),
  );

  return unionSet;
};

const intersection = (listOfSets, properties) => {
  if (!listOfSets) {
    return new Set();
  }

  const sets = listOfSets.map(
    (key) => properties[key]?.cellIds || null,
  ).filter(
    (set) => set && set.size > 0,
  );

  if (sets.length === 0) {
    return new Set();
  }

  const intersectionSet = sets.reduce(
    (acc, curr) => new Set([...acc].filter((x) => curr.has(x))),
  );

  return intersectionSet;
};

const complement = (listOfSets, properties) => {
  if (!listOfSets) {
    return new Set();
  }

  // get the ids of all selected cells
  const selectedCells = listOfSets.map(
    (key) => properties[key]?.cellIds || null,
  ).filter(
    (set) => set && set.size > 0,
  ).reduce(
    (acc, curr) => new Set([...acc, ...curr]),
  );

  // get the ids of all cells in the dataset
  // All cells are assumed to be included in at least 1 cluster
  const complementSet = Object.values(properties).map(
    (cluster) => cluster.cellIds,
  ).filter(
    (set) => set && set.size > 0,
  ).reduce(
    (acc, curr) => new Set([...acc, ...curr]),
  );

  // remove all cells that are selected
  if (selectedCells.size > 0) {
    selectedCells.forEach((x) => { complementSet.delete(x); });
  }

  // return the rest of the cells that are in the dataset and were not selected
  return complementSet;
};

export {
  union,
  intersection,
  complement,
  unionByCellClass,
};
