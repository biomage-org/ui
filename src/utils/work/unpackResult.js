import { decompress } from 'fflate';

const logWithDate = (logStr) => {
  const date = new Date();
  const hour = date.getHours();
  const minutes = date.getMinutes();
  const seconds = date.getSeconds();
  const milliseconds = date.getMilliseconds();

  console.log(
    `[${(hour < 10) ? `0${hour}` : hour
    }:${(minutes < 10) ? `0${minutes}` : minutes
    }:${(seconds < 10) ? `0${seconds}` : seconds
    }.${(`00${milliseconds}`).slice(-3)
    }] ${logStr}`,
  );
};

const unpackResult = async (storageResp, taskName = null) => {
  // SeuratObject can fail to download when loaded into memory
  if (taskName === 'DownloadAnnotSeuratObject') {
    const blob = await storageResp.blob();
    return (blob);
  }

  const arrayBuf = await storageResp.arrayBuffer();

  return decompressUint8Array(new Uint8Array(arrayBuf));
};

const decompressUint8Array = async (arrayBuf) => {
  logWithDate('arrayBufDebug');
  return (
    new Promise((resolve, reject) => {
      // const uint8array = new Uint8Array(arrayBuf);
      const uint8array = arrayBuf;

      decompress(uint8array, (err, decompressed) => {
        logWithDate('decompressedDebug');

        logWithDate('errDebug');
        logWithDate(err);
        if (err) {
          reject(err);
        } else {
          resolve(decompressed);
        }
      });
    })
  );
};

export { decompressUint8Array };
export default unpackResult;
