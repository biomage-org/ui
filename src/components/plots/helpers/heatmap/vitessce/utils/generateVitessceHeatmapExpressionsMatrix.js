import _ from 'lodash';
import { convertRange } from 'utils/plotUtils';

const scaledTo255 = (rowOfExpressions) => {
  const min = _.min(rowOfExpressions);
  const max = _.max(rowOfExpressions);

  return rowOfExpressions.map((value) => convertRange(value, [min, max], [0, 255]));
};

const generateVitessceHeatmapExpressionsMatrix = (cellOrder, geneOrder, expression) => {
  const geneExpressionsDataMatrix = [];

  geneOrder.forEach((gene) => {
    const { matrix } = expression;

    if (!matrix.geneIsLoaded(gene)) return;

    console.log('[DEBUG] - BEGUN matrix.getTruncatedExpression(gene)');
    const truncatedExpression = matrix.getTruncatedExpression(gene);
    console.log('[DEBUG] - FINISHED matrix.getTruncatedExpression(gene)');

    console.log('[DEBUG] - BEGUN cellOrder.map((cellId) => truncatedExpression[cellId])');
    const geneExpressions = cellOrder.map((cellId) => truncatedExpression[cellId]);
    console.log('[DEBUG] - FINISHED cellOrder.map((cellId) => truncatedExpression[cellId])');

    console.log('[DEBUG] - BEGUN scaledTo255');
    const scaledGeneExpressions = scaledTo255(geneExpressions);
    console.log('[DEBUG] - FINISHED scaledTo255');

    geneExpressionsDataMatrix.push(scaledGeneExpressions);
  });

  const cellExpressionsData = _.flatten(_.unzip(geneExpressionsDataMatrix));

  return cellExpressionsData;
};

export default generateVitessceHeatmapExpressionsMatrix;
