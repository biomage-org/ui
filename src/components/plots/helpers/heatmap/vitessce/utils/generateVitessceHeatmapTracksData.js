import getContainingCellSetsProperties from 'utils/cellSets/getContainingCellSetsProperties';
import { hexToRgb } from 'utils/plotUtils';

const generateVitessceHeatmapTracksData = (trackOrder, cellSets, cells) => {
  const colorForCell = (cellId, cellClassKey) => {
    const { color: cellColor = null } = getContainingCellSetsProperties(
      cellId,
      [cellClassKey],
      cellSets,
    )[cellClassKey][0] ?? {};

    return hexToRgb(cellColor) ?? hexToRgb('#f5f8fa');
  };

  const cellIdsColorsMap = new Map();

  cells.forEach((cellId) => {
    const allColorsForCell = trackOrder.map((cellClassKey) => colorForCell(cellId, cellClassKey));

    cellIdsColorsMap.set(`${cellId}`, allColorsForCell);
  });

  return cellIdsColorsMap;
};

export default generateVitessceHeatmapTracksData;
