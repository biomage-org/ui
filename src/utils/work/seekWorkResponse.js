import moment from 'moment';

import getAuthJWT from 'utils/getAuthJWT';
import fetchAPI from 'utils/http/fetchAPI';
import unpackResult from 'utils/work/unpackResult';
import parseResult from 'utils/work/parseResult';
import WorkTimeoutError from 'utils/errors/http/WorkTimeoutError';
import WorkResponseError from 'utils/errors/http/WorkResponseError';
import httpStatusCodes from 'utils/http/httpStatusCodes';

const throwResponseError = (response) => {
  throw new Error(`Error ${response.status}: ${response.text}`, { cause: response });
};

// getRemainingWorkerStartTime returns how many more seconds the worker is expected to
// need to be running with an extra 1 minute for a bit of leeway
const getRemainingWorkerStartTime = (creationTimestamp) => {
  const now = new Date();
  const creationTime = new Date(creationTimestamp);
  const elapsed = parseInt((now - creationTime) / (1000), 10); // gives second difference

  // we assume a worker takes up to 5 minutes to start
  const totalStartup = 5 * 60;
  const remainingTime = totalStartup - elapsed;
  // add an extra minute just in case
  return remainingTime + 60;
};

const seekFromS3 = async (ETag, experimentId, taskName) => {
  let response;
  try {
    const url = `/v2/workResults/${experimentId}/${ETag}`;

    response = await fetchAPI(url);
  } catch (e) {
    if (e.statusCode === httpStatusCodes.NOT_FOUND) {
      return null;
    }

    throw e;
  }

  const storageResp = await fetch(response.signedUrl);

  if (!storageResp.ok) {
    throwResponseError(storageResp);
  }

  const unpackedResult = await unpackResult(storageResp);
  const parsedResult = parseResult(unpackedResult, taskName);

  return parsedResult;
};

// getTimeoutDate returns the date resulting of adding 'timeout' seconds to
// current time.
const getTimeoutDate = (timeout) => moment().add(timeout, 's').toISOString();

const resetTimeout = (id, request, newTimeout, reject) => {
  clearTimeout(id);
  setTimeout(() => {
    reject(new WorkTimeoutError(getTimeoutDate(newTimeout), request));
  }, (newTimeout) * 1000);
};

const dispatchWorkRequest = async (
  experimentId,
  body,
  timeout,
  ETag,
  requestProps,
) => {
  console.error('dispatching work request', body);
  const { default: connectionPromise } = await import('utils/socketConnection');
  const io = await connectionPromise;

  const timeoutDate = getTimeoutDate(timeout);
  const authJWT = await getAuthJWT();

  const request = {
    ETag,
    socketId: io.id,
    experimentId,
    ...(authJWT && { Authorization: `Bearer ${authJWT}` }),
    timeout: timeoutDate,
    body,
    ...requestProps,
  };

  const timeoutPromise = new Promise((resolve, reject) => {
    const id = setTimeout(() => {
      reject(new WorkTimeoutError(timeoutDate, request));
    }, timeout * 1000);

    io.on(`WorkerInfo-${experimentId}`, (res) => {
      const { response: { podInfo: { name, creationTimestamp, phase } } } = res;
      console.log('received worker info: ', res); // TODO: remove
      const extraTime = getRemainingWorkerStartTime(creationTimestamp);

      // this worker info indicates that the work request has been received but the worker
      // is still spinning up so we will add extra time to account for that.
      if (phase === 'Pending' && extraTime > 0) {
        console.log(`WorkerInfo-${experimentId}: ${name} [${creationTimestamp}]: adding ${extraTime} seconds to timeout.`);
        const newTimeout = timeout + extraTime;
        resetTimeout(id, request, newTimeout, reject);
      }
    });

    // this experiment update is received whenever a worker finishes any work request
    // related to the current experiment. We extend the timeout because we know
    // the worker is alive and was working on another request of our experiment //
    // (so this request was in queue)
    io.on(`ExperimentUpdates-${experimentId}`, (res) => {
      const { request: completedRequest } = res;
      console.log('received experiment update: ', completedRequest); // TODO: remove
      console.log(`ExperimentUpdates-${experimentId}: refreshing ${timeout} seconds timeout.`);
      resetTimeout(id, request, timeout, reject);
    });
  });

  const responsePromise = new Promise((resolve, reject) => {
    io.on(`WorkResponse-${ETag}`, (res) => {
      const { response } = res;

      if (response.error) {
        const { errorCode, userMessage } = response;
        console.error(errorCode, userMessage);

        return reject(
          new WorkResponseError(errorCode, userMessage, request),
        );
      }

      // If no error, the response should be ready on S3.
      // In this case, return true
      return resolve();
    });
  });

  const result = Promise.race([timeoutPromise, responsePromise]);

  // TODO switch to using normal WorkRequest for v2 requests
  io.emit('WorkRequest-v2', request);

  return result;
};

export { dispatchWorkRequest, seekFromS3 };
