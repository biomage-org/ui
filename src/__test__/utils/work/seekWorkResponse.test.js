import { v4 as uuidv4 } from 'uuid';
import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';
import SocketMock from 'socket.io-mock';
import { dispatchWorkRequest, seekFromS3 } from 'utils/work/seekWorkResponse';
import fake from '__test__/test-utils/constants';
import mockAPI, { generateDefaultMockAPIResponses } from '__test__/test-utils/mockAPI';

import unpackResult from 'utils/work/unpackResult';
import parseResult from 'utils/work/parseResult';

/**
 * jest.mock calls are automatically hoisted to the top of the javascript
 * during compilation. Accordingly, `mockEmit` and `mockOn` as exported
 * from jest.mock will be accessible under `socketConnectionMocks`, even
 * if they do not appear in the original file.
 */
import * as socketConnectionMocks from 'utils/socketConnection';
import { waitFor } from '@testing-library/react';

enableFetchMocks();
uuidv4.mockImplementation(() => 'my-random-uuid');

jest.mock('uuid');
jest.mock('dayjs', () => () => jest.requireActual('dayjs')('4022-01-01T00:00:00.000Z'));
jest.mock('utils/socketConnection', () => {
  const mockEmit = jest.fn();
  const mockOn = jest.fn();

  return {
    __esModule: true,
    default: new Promise((resolve) => {
      resolve({ emit: mockEmit, on: mockOn, id: '5678' });
    }),
    mockEmit,
    mockOn,
  };
});

jest.mock('utils/work/parseResult');

const mockResponseDataCompressed = 'mockResCompressed';
const mockResponseDataDecompressed = 'mockResDecompressed';

jest.mock('utils/work/unpackResult', () => ({
  __esModule: true,
  default: jest.fn(),
  decompressUint8Array: jest.fn().mockImplementation(() => mockResponseDataDecompressed),
}));

const taskName = 'GetEmbedding';

describe('dispatchWorkRequest unit tests', () => {
  const experimentId = fake.EXPERIMENT_ID;
  const timeout = 30;
  const body = {
    name: taskName,
    type: 'fake task',
  };

  const dispatchMock = jest.fn();

  beforeEach(async () => {
    jest.clearAllMocks();

    fetchMock.resetMocks();
    fetchMock.mockIf(/.*/, mockAPI(generateDefaultMockAPIResponses(fake.EXPERIMENT_ID)));

    const socketMock = new SocketMock();

    parseResult.mockImplementationOnce(() => 'mockParsedResult');

    socketConnectionMocks.mockEmit.mockImplementation((workRequestType, requestBody) => {
      const podInfo = {
        response: {
          podInfo: {
            name: 'worker-pod',
            creationTimestamp: '2022-04-29T07:48:47.000Z',
            phase: 'Pending',
          },
        },
      };

      // mockDecompressUint8Array.mockReturnValue(Promise.resolve(mockResponseDataDecompressed));

      // This is a mocked response emit response from server
      socketMock.socketClient.emit(`WorkResponse-${requestBody.ETag}`, mockResponseDataCompressed);
      socketMock.socketClient.emit(`WorkerInfo-${fake.EXPERIMENT_ID}`, podInfo);
    });

    socketConnectionMocks.mockOn.mockImplementation((channel, socketCallback) => {
      // This is a listener for the response from the server
      socketMock.on(channel, (responseBody) => {
        socketCallback(responseBody);
      });
    });
  });

  it('Sends work to the worker when called', async () => {
    fetchMock.mockResponse(JSON.stringify({ signedUrl: 'http://www.apiUrl:portNum/path/blabla' }));

    const resultPromise = dispatchWorkRequest(
      experimentId, body, timeout, 'facefeed', null, dispatchMock,
    );

    // Wait for subscription to WorkResponse-facefeed
    await waitFor(() => {
      expect(socketConnectionMocks.mockOn.mock.calls.length).toBeGreaterThan(0);

      const [, workResponseHandler] = socketConnectionMocks.mockOn.mock.calls.find(([channel]) => channel === 'WorkResponse-facefeed');

      workResponseHandler(mockResponseDataCompressed);
    });

    const result = await resultPromise;

    expect(result).toEqual({ data: 'mockParsedResult' });

    expect(parseResult).toHaveBeenCalledWith(mockResponseDataDecompressed);

    expect(socketConnectionMocks.mockOn).toHaveBeenCalledTimes(3);
    expect(socketConnectionMocks.mockOn.mock.calls.map(([eventName]) => eventName)).toMatchSnapshot('EventNames');
  });

  it('Returns an error if there is error in the response.', async () => {
    socketConnectionMocks.mockOn.mockImplementation(async (channelName, cb) => {
      let response = null;

      if (channelName.match('WorkerInfo')) {
        response = {
          podInfo: {
            name: 'worker-pod',
            creationTimestamp: '2022-04-29T07:48:47.000Z',
            phase: 'Pending',
          },
        };
      }

      if (channelName.match('WorkResponse')) {
        response = {
          error: true,
          errorCode: 'MOCK_ERROR_CODE',
          userMessage: 'Mock worker error message',
        };
      }

      cb({ response });
    });

    expect(async () => {
      await dispatchWorkRequest(experimentId, body, timeout, 'facefeed', null, dispatchMock);
    }).rejects.toEqual(new Error('MOCK_ERROR_CODE: Mock worker error message'));
  });
});

describe('seekFromS3 unit tests', () => {
  const result = 'someResult';

  const validSignedUrl = 'https://s3.mock/validSignedUrl';
  const invalidSignedUrl = 'https://s3.mock/invalidSignedUrl';
  const notFoundSignedUrl = 'https://s3.mock/notFoundSignedUrl';

  const s3ErrorResponse = new Response('Forbidden', { status: 403 });
  const notFoundErrorResponse = new Response('NotFound', { status: 404 });

  beforeAll(async () => {
    fetchMock.mockIf(/.*/, (req) => {
      const path = req.url;

      if (path.endsWith(validSignedUrl)) return Promise.resolve(result);
      if (path.endsWith(invalidSignedUrl)) return Promise.resolve(s3ErrorResponse);
      if (path.endsWith(notFoundSignedUrl)) return Promise.resolve(notFoundErrorResponse);

      return {
        status: 500,
        body: 'Something eror with test',
      };
    });
  });

  beforeEach(async () => {
    jest.clearAllMocks();
  });

  it('Should return null when response status is not found', async () => {
    const finalResult = await seekFromS3(taskName, notFoundSignedUrl);

    expect(finalResult).toBeNull();
  });

  it('Should return results correctly', async () => {
    unpackResult.mockReturnValueOnce('mockUnpackedResult');
    parseResult.mockReturnValueOnce('mockParsedResult');

    const finalResult = await seekFromS3(taskName, validSignedUrl);

    expect(finalResult).toEqual('mockParsedResult');

    expect(unpackResult).toHaveBeenCalledTimes(1);
    const response = unpackResult.mock.calls[0][0];
    const mockResponsePayload = await response.text();
    expect(mockResponsePayload).toEqual(result);

    expect(parseResult).toHaveBeenCalledWith('mockUnpackedResult', taskName);
  });

  it('Should throw an error if fetching returns an error', async () => {
    await expect(async () => {
      await seekFromS3(taskName, invalidSignedUrl);
    }).rejects.toThrow(new Error(`Error ${s3ErrorResponse.status}: ${s3ErrorResponse.text}`, { cause: s3ErrorResponse }));
  });
});
