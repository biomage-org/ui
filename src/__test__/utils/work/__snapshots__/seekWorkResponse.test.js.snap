// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dispatchWorkRequest unit tests Sends work to the worker when called: EventNames 1`] = `
Array [
  "WorkerInfo-testae48e318dab9a1bd0bexperiment",
  "Heartbeat-testae48e318dab9a1bd0bexperiment",
  "WorkResponse-facefeed",
]
`;
